[Joshen Lim] iP

joshenx opened the pull request:
# Duke Chatbot by joshenx
> This Duke Chatbot was made by @joshenx. Check him out at his [GitHub](github.com/joshenx) or his [Linktree](linktr.ee/joshenx)! :+1:

## Nonsense 
Ever wondered what goes on in the *scary* module CS2102? Check it out here: 
![PostgreSQL](https://i.imgur.com/U3YFDOR.png)

Anyway, check out Duke Chatbot, it has
* cool feature 1
* cool feature 2

and 
1. numbered feature 1
2. numbered feature 2

so it should be.. **really cool**.

So far, the implemented features are:
- [x] Week 1 Features
- [x] Week 2 Features
- [x] Week 3 Features
- [ ] Week 4 Features

## Running The Program
Try it out with the `main` method in Main.java!

It should look like this:
```
public class Main {
    public static void main(String[] args) throws IOException {
        new Duke("data/tasks.txt").run();
    }
}
```

Commit:
Add Gradle support

joshenx committed:
Add Level-1

joshenx committed:
Add gitignore files and Level-2

joshenx committed:
Add Level-3 and Task class

joshenx committed:
added Deadline, Event, Todo classes and added Level-4

joshenx committed:
fixed typo in Event.java and added Text Ui Testing

joshenx committed:
Add Level-5 and updated io test

joshenx committed:
add Level-6, update io tests

joshenx committed:
implement enums

joshenx committed:
Add io/save feature for Level-7

joshenx committed:
Add time/date parsing for Level-8

joshenx committed:
Fix Level-7 and Level-8 merge conflicts

joshenx committed:
Abstract most methods

joshenx committed:
Divide classes into packages and put static constants in Formats class

joshenx committed:
Fix 1 bug

joshenx committed:
Add JUnit tests for Deadline and Todo classes

joshenx committed:
Add jar file

joshenx committed:
Add Javadocs

joshenx committed:
Modify code to follow coding standards

joshenx committed:
Add Find tasks feature

joshenx committed:
Merge Coding Standard changes with master

joshenx committed:
Merge branch 'branch-Level-9'

joshenx committed:
Merge find feature with master

joshenx committed:
Merge branch 'add-gradle-support'

joshenx committed:
Add gradle support

joshenx committed:
Add CheckStyle features to new branch

caipng commented on .gitignore:
```diff
@@ -13,5 +13,10 @@ src/main/resources/docs/
 *.iml
 bin/
 
+# Files that should not be version controlled
```
Maybe a more explanatory comment such as
```suggestion
# Java class and package files
```

caipng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+    private Ui ui;
+    private Storage storage;
+
+    /**
+     * @param filePath to read/write task list data
+     * @throws IOException
+     */
```
There are some violations of the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html#comments) with respect to the Javadoc form.

caipng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+    public void run() {
+        boolean isActive = true;
+
+        String welcomeMsg = "Hey, I'm duke.Duke.\n"
+                + "What's up?\n";
+        String exitMsg = "Bye! Hope I helped!\n"
+                + "See you next time :)\n";
+
+        System.out.println(welcomeMsg);
+        System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
```
It may be useful to abstract out these messages as class constants.

Something like this for example:
```java
private static final String MSG_WELCOME = "Hey, I'm duke.Duke.\nWhat's up?\n";
private static final String MSG_EXIT = "Bye! Hope I helped!\nSee you next time :)\n";
```

joshenx replied:
Good suggestion, thank you!

caipng commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+                isActive = c.isActive();
+
+            } catch (DukeException e) {
+                System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
+                System.out.println(e.getMessage() + "\n");
+                System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
+            } catch (DateTimeParseException e) {
+                System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
+                System.out.println("Please enter the date and time "
+                        + "in dd/mm/yyyy hh:mma format! :\n)");
+                System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
+            }
+        }
+
+        System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
+        System.out.println(exitMsg);
+        System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
```
There is some code repetition here. How about an abstraction such as
```java
void say(String message) {
    System.out.println(MSG_DIVIDER);
    System.out.println(message);
    System.out.println(MSG_DIVIDER);
}
```
Then, for example, you can do
```java
say(MSG_EXIT);
say(e.getMessage());
```

caipng commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
+        System.out.println("Alright! I added this task: \n" + toAdd.toString() + "\n");
+        tasks.add(toAdd);
```
It may be better to swap these 2 lines as the message implies that the task has been successfully added (in case  something goes wrong in `tasks.add(toAdd)`)

caipng commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,51 @@
+     * @param list list of tasks
+     * @return ArrayList in String format
+     */
+    public static String arrayToString(ArrayList<Task> list) {
```
A better function name may be
```suggestion
    public static String taskListToString(ArrayList<Task> list) {
```

Even better may be to override the `toString()` method of the `TaskList` class.

yulonglim replied:
I agree with the comment!

caipng commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,110 @@
+            lines = Files.readAllLines(Paths.get(fileName));
+
+            for (String line : lines) {
+                //System.out.println(line);
```
```suggestion
```

caipng commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,110 @@
+                outputWriter.newLine();
+            }
+            outputWriter.flush();
+            outputWriter.close();
```
Should this be in the `finally` clause instead?

joshenx replied:
That would make sense!

caipng commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,39 @@
+ * Represents a task for the user to do.
+ */
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
Maybe these fields can be private?

caipng reviewed (commented):
Overall good effort on the implementation! However, there are some code quality issues here and there. Please do bring up any questions you may have regarding my comments.

joshenx committed:
Add GUI for Duke

joshenx committed:
Fix bugs and add varargs

shaliniseshadri commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,95 @@
+    private Ui ui;
+    private Storage storage;
+
+    /**
```
There is a missing description here in the Javadocs. Maybe you could add a one-liner about the usage of each constructor?

shaliniseshadri commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,60 @@
+    public Command() {
+    }
+
+    /**
+     * Executes the command and returns the result.
+     */
```
Maybe you could include the javadocs for the params here?

shaliniseshadri commented on src/test/java/DeadlineTest.java:
```diff
@@ -0,0 +1,25 @@
+
+public class DeadlineTest {
+    @Test
+    public void outputTest1() {
```
Minor edit: Maybe you can consider changing the name of the test methods to match the format just to make it a little clearer:  
featureUnderTest_testScenario_expectedBehavior()
Source: https://se-education.org/guides/conventions/java/intermediate.html#naming

shaliniseshadri commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,22 @@
+
+    }
+
+    public static void showLoadingError() {
```
Can consider removing the empty methods if you aren't/don't plan on using them.

shaliniseshadri reviewed (commented):
Overall, the code styling looks fine. Just suggested a few minor edits which I thought might be appropriate. Please feel free to clarify/correct me if I made a wrong assumption in any comment.

yulonglim commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,12 @@
+ * Entry point of the Duke application.
+ */
+public class Launcher {
+    public static void main(String[] args)  {
```
There's is an extra space after the ).

yulonglim commented on src/main/java/duke/commands/CompleteCommand.java:
```diff
@@ -22,8 +22,8 @@ public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException
         } else {
             tasks.get(index).markAsDone();
             System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
```
Bot have been shifted to GUI version, maybe dont require print statements anymore?

yulonglim commented on src/main/java/duke/commands/CompleteCommand.java:
```diff
@@ -22,8 +22,8 @@ public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException
-                    tasks.get(index).toString() + "\n");
+            System.out.println("Awesome! I marked this as done:\n"
+                    + tasks.get(index).toString() + "\n");
             System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
```
Maybe better to keep repeated printed items like the long lines as constants?

yulonglim commented on src/main/java/duke/DukeException.java:
```diff
@@ -3,14 +3,14 @@
 public class DukeException extends Exception {
     private static final String INVALID_COMMAND = "☹ OOPS!!! I'm sorry, but I don't know what that means :-(";
     private static final String EMPTY_DESC = "☹ OOPS!!! The description of a task cannot be empty.";
```
Maybe better to write out DESC in full?

yulonglim commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -47,6 +53,13 @@ public static CommandEnum valueOfLabel(String label) {
+     * @throws DukeException if the command executed has errors
+     */
     public static Command parse(String fullCommand) throws DukeException {
         String[] tokens = fullCommand.split("\\s+", 2);
```
tokens can be confusing, maybe try words? (wasn't intuitive to me what was the variable for until i saw the code however i feel that it is just a tiny change you might want to consider!)

yulonglim reviewed (commented):
Code is extremely well written, had a hard time finding even small suggestions.

joshenx committed:
Add array-converting functionality to Ui class

Ui class does not have any function.

Methods that assist functionality (such as converting an array to string) should be handled by the Ui class.

Let's bring the tasklistToString() method from the Command class to the Ui class. This will follow OOP principles better.

joshenx committed:
Add assertions at important points

joshenx committed:
Improve code quality of Duke application.

There is a lot of commented chunks of code, and the code does not pass checkstyle tests. Some variables are named incorrectly.

Code quality should be improved nearing submission of iP.

Let's improve on OOP principles, use better naming standards and fix coding standard violations.

joshenx committed:
Merge pull request #1 from joshenx/branch-A-Assertions

Add assertions at important points

joshenx committed:
Merge Assertions and Code Quality changes

joshenx committed:
Merge pull request #2 from joshenx/branch-A-CodeQuality

Improve code quality of Duke application

joshenx committed:
Add archive feature

joshenx committed:
Improve the GUI

joshenx committed:
Add User Guide

joshenx committed:
Set theme jekyll-theme-slate

joshenx committed:
Fix user guide

joshenx committed:
Merge branch 'master' of https://github.com/joshenx/ip

joshenx committed:
Remove dead code
